### PR TITLE
Integrate changes to support credential generators

### PR DESCRIPTION
**Pull request by BrunoCoimbra**
This pull request integrates the resulting code of the GWMS hackathon on March 18th.

I made improvements to the handling of credential plugins on the frontend main loop, and a couple of small fixes to the scitoken_callout plugin. I also changed our spec file to install plugins under the `plugins` directory at `/etc/gwms-frontend/plugin.d`.

### CODE REVIEW

File: frontend/glideinFrontendElement.py [hunk 1]
```diff
@@ -982,6 +968,108 @@ def iterate_one(self):
...
+            cred_type_map = eval(cred_type_data)
+            trust_domain_map = eval(trust_domain_data)
+            for cfname in cred_type_map:
+                if cred_type_map[cfname] == "scitoken":
+                    if trust_domain_map[cfname] == trust_domain:
+                        scitoken_fullpath = cfname
+
+        if os.path.exists(scitoken_fullpath):
+            try:
+                logSupport.log.debug("found scitoken %s" % scitoken_fullpath)
```
Comment (namrathaurs) [inline frontend/glideinFrontendElement.py:999]: I see that logSupport is being used here to print the string. Does f-string formatting apply here like how we use it in python 3.x?

File: frontend/glideinFrontendElement.py [hunk 2]
```diff
@@ -982,6 +968,108 @@ def iterate_one(self):
...
+                stkn = ""
+                with open(scitoken_fullpath) as fbuf:
+                    for line in fbuf:
+                        stkn += line
+                stkn = stkn.strip()
+                return stkn
+            except Exception as err:
+                logSupport.log.exception("failed to read scitoken: %s" % err)
```
Comment (namrathaurs) [inline frontend/glideinFrontendElement.py:1007]: same comment as previous

File: frontend/glideinFrontendElement.py [hunk 3]
```diff
@@ -982,6 +968,108 @@ def iterate_one(self):
...
+            generators = elementDescript.frontend_data.get("CredentialGenerators")
+        if not trust_domain_data:
+            trust_domain_data = elementDescript.frontend_data.get("ProxyTrustDomains")
+        if trust_domain_data and generators:
+            generators_map = eval(generators)
+            trust_domain_map = eval(trust_domain_data)
+            for cfname in generators_map:
+                if trust_domain_map[cfname] == trust_domain:
+                    generator = generators_map[cfname]
+                    logSupport.log.debug("found credential generator plugin %s" % generator)
```
Comment (namrathaurs) [inline frontend/glideinFrontendElement.py:1056]: same comment as previous

File: frontend/glideinFrontendElement.py [hunk 4]
```diff
@@ -982,6 +968,108 @@ def iterate_one(self):
...
+                        if not generator in plugins:
+                            plugins[generator] = import_module(generator)
+                        entry = {
+                            "name": glidein_el["attrs"].get("EntryName"),
+                            "gatekeeper": glidein_el["attrs"].get("GLIDEIN_Gatekeeper"),
+                        }
+                        stkn, _ = plugins[generator].get_credential(logSupport, group_name, entry, trust_domain)
+                        return stkn
+                    except ModuleNotFoundError:
+                        logSupport.log.warning("Failed to load credential generator plugin %s" % generator)
```
Comment (namrathaurs) [inline frontend/glideinFrontendElement.py:1067]: same comment as previous

File: frontend/glideinFrontendElement.py [hunk 5]
```diff
@@ -982,6 +968,108 @@ def iterate_one(self):
...
+                    except Exception as e:  # catch any exception from the plugin to prevent the frontend from crashing
+                        logSupport.log.warning("Failed to generate credential: %s." % e)
```
Comment (namrathaurs) [inline frontend/glideinFrontendElement.py:1069]: same comment as previous

File: plugins/scitokens_callout.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    tkn_str = ""
+    tkn_lifetime = tkn_max_lifetime
+    tmpnm = ""
+
+    audience = None
+    if "gatekeeper" in entry:
+        audience = entry["gatekeeper"].split()[-1]
+    subject = None
+    if "name" in entry:
+        subject = "vofrontend-%s" % (entry["name"])
```
Comment (namrathaurs) [inline plugins/scitokens_callout.py:48]: is f-string formatting possible here?

File: plugins/scitokens_callout.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+    tkn_dir = "/var/lib/gwms-frontend/tokens.d"
```
Comment (namrathaurs) [inline plugins/scitokens_callout.py:50]: any possibility to avoid the hardcoding?

File: plugins/scitokens_callout.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                f' aud="{audience}"'
+                f' scope="{scope}"'
+                f' wlcg.ver="{wlcg_ver}"'
+            )
+            tkn_str = subprocessSupport.iexe_cmd(cmd)
+            os.write(fd, tkn_str.encode("utf-8"))
+            os.close(fd)
+            shutil.move(tmpnm, tkn_file)
+            os.chmod(tkn_file, 0o600)
+            logger.debug("created token %s" % tkn_file)
```
Comment (namrathaurs) [inline plugins/scitokens_callout.py:80]: I see that logger is being used here to print the string. Does f-string formatting apply here like how we use it in python 3.x?

File: plugins/scitokens_callout.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        elif os.path.exists(tkn_file):
+            with open(tkn_file) as fbuf:
+                for line in fbuf:
+                    tkn_str += line
+            tkn_str = tkn_str.strip()
+            tkn_lifetime = tkn_max_lifetime - tkn_age
+    except Exception as err:
+        logger.log.warning("failed to create %s" % tkn_file)
```
Comment (namrathaurs) [inline plugins/scitokens_callout.py:88]: same comment as previous

**Review by namrathaurs — CHANGES_REQUESTED** [review]
Left a few comments as part of my review although not sure if they are really applicable. Since the patch resulting from the hackathon is relatively new to me, it might also be nice if @ddbox or @mmascher can do a quick review (those files that have major changes) to make sure I haven't missed anything important to raise.

**Review by mambelli — CHANGES_REQUESTED** [review]

File: frontend/glideinFrontendElement.py [hunk 10]
```diff
@@ -982,6 +968,108 @@ def iterate_one(self):
...
+        if not generators:
+            generators = elementDescript.frontend_data.get("CredentialGenerators")
+        if not trust_domain_data:
+            trust_domain_data = elementDescript.frontend_data.get("ProxyTrustDomains")
+        if trust_domain_data and generators:
+            generators_map = eval(generators)
+            trust_domain_map = eval(trust_domain_data)
+            for cfname in generators_map:
+                if trust_domain_map[cfname] == trust_domain:
+                    generator = generators_map[cfname]
```
[the remark this answers is not in the document]
Comment (BrunoCoimbra) [reply frontend/glideinFrontendElement.py:1055]: This shouldn't be a problem, as earlier in the code, we source the `generators` variable from the `CredentialGenerators` group or frontend config. If no generator is provided, the generators variable is `None` and will abort the function.

Here's the relevant portion of the code:
```python
generators = elementDescript.element_data.get("CredentialGenerators")
trust_domain_data = elementDescript.element_data.get("ProxyTrustDomains")
if not generators:
    generators = elementDescript.frontend_data.get("CredentialGenerators")
if not trust_domain_data:
    trust_domain_data = elementDescript.frontend_data.get("ProxyTrustDomains")
if trust_domain_data and generators:
    ...
```

**Review by mambelli — APPROVED** [review]

**Pull request merged**